src/main/java/org/cbioportal/cdd/service/internal/CDDServiceMskVocabularyImpl.java

```diff
@@ -0,0 +1,132 @@
+            throw new ClinicalMetadataSourceUnresponsiveException("MskVocabulary system did not return valid set of terms");
+        }
+        for (MskVocabularyField mskVocabularyField : mskVocabularyList) {
+            ClinicalAttributeMetadata clinicalAttributeMetadata = mskVocabularyConverter.convertToClinicalAttributeMetadata(mskVocabularyField);
+            clinicalAttributeMetadataCache.put(clinicalAttributeMetadata.getColumnHeader(), clinicalAttributeMetadata);
+        }
+    }
+
+    @Override
+    public List<ClinicalAttributeMetadata> getClinicalAttributeMetadata(String cancerStudy)
+        throws ClinicalMetadataSourceUnresponsiveException, CancerStudyNotFoundException {
+        if (clinicalAttributeMetadataCache.isEmpty()) {
+            fillClinicalAttributeMetadataCache();
+        }
+        return new ArrayList<ClinicalAttributeMetadata>(clinicalAttributeMetadataCache.values());
+    }
+
+    @Override
+    public List<ClinicalAttributeMetadata> getMetadataByColumnHeaders(String cancerStudy, List<String> columnHeaders)
+        throws ClinicalAttributeNotFoundException, ClinicalMetadataSourceUnresponsiveException, CancerStudyNotFoundException {
+        ArrayList<ClinicalAttributeMetadata> clinicalAttributeMetadata = new ArrayList<ClinicalAttributeMetadata>();
+        ArrayList<String> invalidColumnHeaders = new ArrayList<String>();
+        if (clinicalAttributeMetadataCache.isEmpty()) {
+            fillClinicalAttributeMetadataCache();
+        }
+        for (String columnHeader : columnHeaders) {
+            try {
+                clinicalAttributeMetadata.add(getMetadataByColumnHeader(columnHeader));
+            } catch (ClinicalAttributeNotFoundException e) {
+                invalidColumnHeaders.add(columnHeader);
+            }
+        }
+        if (invalidColumnHeaders.size() > 0) {
+            throw new ClinicalAttributeNotFoundException(invalidColumnHeaders);
+        }
+        return clinicalAttributeMetadata;
+    }
+
+    @Override
+    public List<ClinicalAttributeMetadata> getMetadataBySearchTerms(List<String> searchTerms, String attributeType, boolean inclusiveSearch)
+        throws ClinicalAttributeNotFoundException, ClinicalMetadataSourceUnresponsiveException {
+        throw new UnsupportedOperationException("search for similar terms within the MSK Standard Vocabulary is not yet implemented");
+    }
+
+    @Override
+    public ClinicalAttributeMetadata getMetadataByColumnHeader(String cancerStudy, String columnHeader)
+        throws ClinicalAttributeNotFoundException, ClinicalMetadataSourceUnresponsiveException, CancerStudyNotFoundException {
+        if (clinicalAttributeMetadataCache.isEmpty()) {
+            fillClinicalAttributeMetadataCache();
+        }
+        return getMetadataByColumnHeader(columnHeader);
+    }
+
+    @Override
+    public List<CancerStudy> getCancerStudies() throws ClinicalMetadataSourceUnresponsiveException {
```

Comment:
note : currently, there is nothing calling this function. This is here for future use (after retiring knowledgeSystems vocabulary perhaps)